### PR TITLE
🎉 Source Typeform: Increase unit test coverage

### LINKED ISSUE #12220: Test Coverage
[Unit Test Files](https://github.com/airbytehq/airbyte/tree/master/airbyte-integrations/connectors/source-typeform/unit_tests) - Coverage seems to be limited, needs more triage.

### PR DESCRIPTION
**Pull request by itaseskii**
## What
Increases unit test coverage for the existing Typeform streams

Closes #12220.

## How
Using pytest

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-typeform/source_typeform/source.py [hunk 1]
```diff
@@ -204,20 +204,20 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
             form_ids = config.get("form_ids", []).copy()
             # verify if form inputted by user is valid
             try:
-                url = f"{TypeformStream.url_base}/me"
+                url = f"{TypeformStream.url_base}me"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-typeform/source_typeform/source.py:207]: maybe use `urljoin` function to ensure this won't break in any case

File: airbyte-integrations/connectors/source-typeform/source_typeform/source.py [hunk 2]
```diff
@@ -204,20 +204,20 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
...
                 auth_headers = {"Authorization": f"Bearer {config['token']}"}
                 session = requests.get(url, headers=auth_headers)
                 session.raise_for_status()
-            except requests.exceptions.BaseHTTPError as e:
+            except Exception as e:
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-typeform/source_typeform/source.py:211]: Why are you changing the exception type error to the generic one?

File: airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+from source_typeform import SourceTypeform
+from source_typeform.source import TypeformStream
+
+logger = AirbyteLogger()
+
+TYPEFORM_BASE_URL = TypeformStream.url_base
+
+
+def test_check_connection(requests_mock, config, empty_response_ok):
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py:11]:
```suggestion
def test_check_connection_success(requests_mock, config, empty_response_ok):
```

File: airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    assert ok and not error
+
+
+def test_check_connection_bad_request_me(requests_mock, config, empty_response_bad):
+    requests_mock.register_uri("GET", TYPEFORM_BASE_URL + "me", empty_response_bad)
+
+    ok, error = SourceTypeform().check_connection(logger, config)
+
+    assert not ok and error
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py:26]: Can you capture the correct output/error and check it? Maybe use pytest match feature

File: airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def test_check_connection_bad_request_forms(requests_mock, config, empty_response_ok, empty_response_bad):
+    requests_mock.register_uri("GET", TYPEFORM_BASE_URL + "me", empty_response_ok)
+    requests_mock.register_uri("GET", TYPEFORM_BASE_URL + "forms/u6nXL7", empty_response_bad)
+
+    ok, error = SourceTypeform().check_connection(logger, config)
+
+    assert not ok and error
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py:35]: same here.

File: airbyte-integrations/connectors/source-typeform/unit_tests/test_streams.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def merge_records(stream: TypeformStream, sync_mode: SyncMode) -> List[Mapping[str, Any]]:
+    merged_records = []
+    for stream_slice in stream.stream_slices(sync_mode=sync_mode):
+        records = stream.read_records(sync_mode=sync_mode, stream_slice=stream_slice)
+        for record in records:
+            merged_records.append(record)
+
+    return merged_records
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-typeform/unit_tests/test_streams.py:67]: move this function to the top

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
some comments thanks @itaseskii

File: airbyte-integrations/connectors/source-typeform/source_typeform/source.py (hunk 1 above)
Comment (itaseskii) [reply airbyte-integrations/connectors/source-typeform/source_typeform/source.py]: done

File: airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py (hunk 3 above)
Comment (itaseskii) [reply airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py]: changed

File: airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py (hunk 4 above)
Comment (itaseskii) [reply airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py]: matched using re.match()

File: airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py (hunk 5 above)
Comment (itaseskii) [reply airbyte-integrations/connectors/source-typeform/unit_tests/test_source.py]: same solution as above

File: airbyte-integrations/connectors/source-typeform/unit_tests/test_streams.py (hunk 6 above)
Comment (itaseskii) [reply airbyte-integrations/connectors/source-typeform/unit_tests/test_streams.py]: moved

File: airbyte-integrations/connectors/source-typeform/source_typeform/source.py (hunk 2 above)
Comment (itaseskii) [reply airbyte-integrations/connectors/source-typeform/source_typeform/source.py:211]: in this case resolving **_config['token']_** if **_token_** is not present was not being handled and the exception was being propagated further down the call stack. it is also the approach that connectors such as Stripe, Mixpanel, etc are using.

**Review by marcosmarxm — APPROVED** [review]
thanks @itaseskii

**Pull request merged**